File: Prolog/Excersice/99-prolog/q9.pl
pack([], []).
pack([X], [X]).

pack([H1 | [H2 | T]], R):-
    H1 = H2,
    pack([H1 | T], R1),
    R = [H2 | R1].

pack([H1 | [H2 | T]], R):-
    H1 \= H2,
    pack([H2 | T], R1),
    R = [H1 | R1].

File: Haskell/Practice/q9.hs
pack l = packHelper [] "" l

packHelper c curr [] = c ++ [curr] 
packHelper c curr [x] = c ++ [curr ++ [x]]
packHelper c curr (x1:x2:xs)  = if x1 == x2 then packHelper c (curr ++ [x1]) (x2:xs) else packHelper (c ++ [curr ++ [x1]]) "" (x2:xs)